### DISCUSSION TITLE
Add possibility to specify id inserting object from webService

### DISCUSSION THREAD

**Discussion by APerricone** [PrestaShop/PrestaShop#33401] [Ideas]
#### Is your feature request related to a problem?

I am adding to our shop management desktop program to prestashop eCommerce, and I want use our id for categories and products.

#### Describe the solution you'd like

I would like a web interface similar to CSV upload, in that case you can specify the id, and the element are added or updated.

#### Alternatives you've considered

Another possibility is specify the id on post or add the id_backoffice or id_desktop or another name.

#### Additional context

none

[a comment by an automated account is not shown]

**Comment by florine2623** [comment]
Hi @APerricone ,
Could you specify exactly what info/fields the feature might need ?
What are `id_backoffice` and `id_desktop` meant to do in that case ?
Thanks!

**Comment by APerricone** [comment]
We have a program, used by a thousand of customers, this program is used to manage products, providers, invoices, cash register, weight scales (the kind that create a sticker with the price), electronic labels, etc etc
We call this kind of software "backoffice", our product is called TLPosWin, and it is used only on Italy on grocery stores.
In this program is possible send the products and the categories to an eCommerce, we already support 4 types of ecommerce, with files or web services.
My problem to support PrestaShop is match the current products/categories with the one on our system. 
An option is use our id as id of categories and product inside presta shop, using the webservice, but it is not possible at the moment, I did a pull request.
Another option is create a CSV with our id, but in this case is not possible upload it automatically.
Another option is save our id as field on categories and products, using the webservice, but add a column is a custom modification and we can not ask it to our customers.
Another option is save prestashop's ids in our system, that is what I am implementing. 
I listed the options from best to worst, I saw request  about force id on webservice on forum and on stackoverflow, that's why i did a pull request.

**Comment by matks** [comment]
> Another option is save our id as field on categories and products, using the webservice, but add a column is a custom modification and we can not ask it to our customers.

This looks like the cleaner way in my opinion. Coupling IDs between your system and a prestashop creates a complex situation where you face multiples issues
- when ID is deleted/updated on one side you need to synchronize data
- if IDs are generated in different manners

Most ERPs I have seen always handle this in new property.
For example if e-commerce system Z has a product object, and 2 ERP A and B are connected to it, I expect product object to have
- column `id` for e-commerce Z
- column `erp_A_id` for ERP A
- column `erp_B_id` for ERP B

This can be done either on prestashop side or on the ERP side. One of them must have this matching "ID of X is equivalent of ID of Y"

> but add a column is a custom modification and we can not ask it to our customers.

Yes you can !
You must create a module that you require your customers to install. Modules can do pretty much whatever they want into PrestaShop, including MySQL schema.
So your module can either add a new column into the `product` SQL table (efficient but dirty) or create a new dedicated SQL table (let's call it `ps_product_extradata_TLPosWin`) with the data you want to store.

**Comment by matks** [comment]
@APerricone Other idea: you can do pretty much anything into a module ... including exposing a webservice ! So you can embed into your module a small import/export webservice built by you that you can use to modify the data you need.

So your way of doing things would be "Hi ! You want to use TLPosWin with PrestaShop ? Download our module for PrestaShop, input in some configuration page your API key and you're done !"

**Comment by APerricone** [comment]
Many thank, but my knowledge of php is not good enought to do this...

**Comment by hibatallahAouadni** [comment]
Ping @Progi1984 & @matks so what the status of this issue? is it **`TBS`** or **`To Do`** ?

**Comment by KminekMatej** [comment]
Ive been facing the same situation in our ERP system. I created the same thing that @matks suggested, table with matching
`localId <=> prestashopId`
From my point of view, this is much clearer solution, than implementing something like `externalId` on prestashop. For instance, now we can synchronize our products (stored in ERP) with multiple prestashop instances (when every dealer has its own prestashop e-commerce solution)

**Comment by hibatallahAouadni** [comment]
Thanks @KminekMatej for your suggestion.

Ping @PrestaShop/prestashop-core-developers do you approve @matks and @KminekMatej solution?

**Comment by tswfi** [comment]
The same id matching problem comes with all object types (categories, attributes etc) I have written a module that adds new table to keep information about objects external id's but Im having some problems with the setter method not being called when doing post calls (put works fine) anyone seen this on their implementations?
